### ISSUE TITLE
Stress-ng sigsuspend test failing with IPC error -13 error on CentOS and Ubuntu

### ISSUE THREAD

**Issue by jinengandhi-intel** [gramineproject/gramine#43]
## Description of the problem

Error message on RHEL:
```
[P1:shim] error: IPC worker: DkStreamWaitForClient failed: -13
[P493:T3110:stress-ng] error: process creation failed
[P496:T3111:stress-ng] error: process creation failed
[P488:T3109:stress-ng] error: process creation failed
```

Error message on Ubuntu:

```
[P1:shim] error: IPC worker: error running IPC callback 5: -13
[::] error: Error during shim_init() in receive_checkpoint_and_restore (-61)
rasp@rasp-WHITLEY:~/validation/oscarlab/Examples/stress-ng$ [P498:T3111:stress-ng] error: process creation failed
[P489:T3109:stress-ng] error: process creation failed
[P483:T3108:stress-ng] error: process creation failed
```

The test was executed with PR 2595 as well as PR 2597 but the issue is not resolved.

The memory used after the test shoots up by 40GB and doesn't free automatically.

Log files attached:
[sigsuspend_rhel.txt](https://github.com/gramineproject/gramine/files/7166907/sigsuspend_rhel.txt)
[sigsuspend_ubuntu.txt](https://github.com/gramineproject/gramine/files/7166908/sigsuspend_ubuntu.txt)

## Steps to reproduce

`graphene-direct stress-ng --sigsuspend 0 --timeout 60s --verbose`

**Comment by mkow** [comment]
> The memory used after the test shoots up by 40GB and doesn't free automatically.

What do you mean by this? That there's a leak in the kernel itself?

**Comment by jinengandhi-intel** [comment]
What I mean is after the test is run with Graphene, since the test is exiting abruptly, it leaves a lot of spawned processes consuming so much memory. This is not seen with Linux native

**Comment by dimakuv** [comment]
Debugged this issue.

Turned out it was Linux (mis)configuration. In particular, the soft limit on the number of open file descriptors per process was too low -- 1024. Please see more details e.g. here: https://muhammadtriwibowo.medium.com/set-permanently-ulimit-n-open-files-in-ubuntu-4d61064429a

On this machine it was:
```
$ ulimit -Sa
...blablabla
open files                      (-n) 1024
```

I set the limit to some higher number: `ulimit -n 65535`

Now I see that the limit changed:
```
$ ulimit -Sa
...blablabla
open files                      (-n) 65535
```

And now `stress-ng sigsuspend` works:
```
$ graphene-direct stress-ng --sigsuspend 128 --timeout 60s --verbose
stress-ng: debug: [1] 144 processors online, 144 processors configured
stress-ng: info:  [1] dispatching hogs: 128 sigsuspend
stress-ng: debug: [1] cache allocate: default cache size: 55296K
stress-ng: debug: [1] starting stressors
stress-ng: debug: [2] stress-ng-sigsuspend: started [2] (instance 0)
... blablabla ...
stress-ng: debug: [4161] stress-ng-sigsuspend: started [4161] (instance 127)
stress-ng: debug: [1] 128 stressors started
stress-ng: debug: [2] stress-ng-sigsuspend: exited [2] (instance 0)
... blablabla ...
stress-ng: debug: [4161] stress-ng-sigsuspend: exited [4161] (instance 127)
stress-ng: debug: [1] process [4161] terminated
stress-ng: info:  [1] successful run completed in 72.16s (1 min, 12.16 secs)
stress-ng: debug: [1] metrics-check: all stressor metrics validated and sane
```

## Analysis

`stress-ng sigsuspend` does this (taken from the manual):
```
#   start N workers that each spawn off 4 child processes that  wait
#   for  a  SIGUSR1  signal from the parent using sigsuspend(2). The
#   parent sends SIGUSR1 signals to each child in rapid  succession.
#   Each sigsuspend wakeup is counted as one bogo operation.
#
sigsuspend 0            # 0 means 1 stressor per CPU
```

I manually verified that `stress-ng sigsuspend`  starts failing with `graphene-direct stress-ng --sigsuspend 101` (101 stressors). So it got me to suspect that there is some limit imposed on the number of processors or some other resources.

After some debugging and straceing, it became obvious that `accept4()` host syscall fails with `-EMFILE` on the IPC worker thread -- when the IPC worker of the main process tries to establish a pipe to the newly spawned child. The IPC worker thread complains with smth like `IPC worker: DkStreamWaitForClient failed` (which is not extremely helpful) and kills the main process. Thus, all child processes just become zombies and keep living forever...

Thinking about this, it makes sense. We have `101` stressors, which gives us `1 + 101*5 = 506` processes (our main process, then 101 stressor processes, each with 4 child processes).

Our main process needs to communicate with these `505` processes. I don't remember why, but apparently there are 2 pipes created between the main process and each child process, which pushes the total number of open pipes in the main process to `505 * 2 = 1010`. At this point, the main process is dangereously close to the Linux-imposed limit of `1024`. Given that the main process needs to open some files, this number spills over this limit, and Linux refuses to serve the main Gramine process. BOOM! The main Gramine process cannot recover from this, and all child Gramine processes become zombies.

**Comment by dimakuv** [comment]
TLDR: Gramine failed because of the Linux soft limit on the number of FDs per process (1024 by default). Increasing this limit to smth big makes Gramine work. See e.g. https://unix.stackexchange.com/questions/84227/limits-on-the-number-of-file-descriptors.

This is not a Gramine bug, but more of a "configure Linux limits to account for Gramine quirks".

The original `stress-ng sigsuspend` works because it doesn't have a need to create the pipes between the main process and all child processes (because the Linux kernel does it already). Gramine can't afford to trust the Linux kernel to deal with pipes, so Gramine has to create and track all these pipes.
